File: hub.go
package main

import (
	"encoding/json"
	"sync"

	"github.com/rs/zerolog/log"
)

// Hub manages rooms and global state.
type Hub struct {
	mu    sync.RWMutex
	rooms map[string]*Room

	storage *Storage
}

func newHub(storage *Storage) *Hub {
	h := &Hub{
		rooms:   make(map[string]*Room),
		storage: storage,
	}

	// Create default room
	defaultRoom := newRoom("default", h, mapWidth, mapHeight)
	h.rooms["default"] = defaultRoom

	// Place whiteboards near map center for visibility
	// Positions snap to tile boundaries for grid alignment (2x3 tile sprites)
	defaultRoom.addObject(&InteractiveObject{
		ID:   "wb-1",
		Type: "whiteboard",
		X:    float64(28 * tileSize),
		Y:    float64(24 * tileSize),
	})
	defaultRoom.addObject(&InteractiveObject{
		ID:   "wb-2",
		Type: "whiteboard",
		X:    float64(34 * tileSize),
		Y:    float64(24 * tileSize),
	})

	// Place a regional chat zone
	// ward-stone is 32x48 (2x3 tiles), origin(0.5, 1) → position at tile boundary
	defaultRoom.addObject(&InteractiveObject{
		ID:    "rc-1",
		Type:  "regional_chat",
		X:     float64(21 * tileSize),
		Y:     float64(16 * tileSize),
		State: json.RawMessage(`{"name":"결계석","radius":80,"retainHistory":false}`),
	})

	return h
}

func (h *Hub) defaultRoom() *Room {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return h.rooms["default"]
}

func (h *Hub) run() {
	h.mu.RLock()
	for _, room := range h.rooms {
		go room.run()
	}
	h.mu.RUnlock()
}

func (h *Hub) closeAll() {
	h.mu.RLock()
	for _, room := range h.rooms {
		room.closeAll()
	}
	h.mu.RUnlock()
	log.Info().Msg("[portal-space] all connections closed")
}

func (h *Hub) wait() {
	h.mu.RLock()
	for _, room := range h.rooms {
		room.wait()
	}
	h.mu.RUnlock()
}

File: frontend/src/server/hub.ts
import type { InteractiveObject, RegionalChatState } from '$lib/types'
import type { Storage } from './storage'
import { Room } from './room'
import { TILE_SIZE, MAP_WIDTH, MAP_HEIGHT } from './protocol'

export class Hub {
	rooms = new Map<string, Room>()
	storage: Storage

	constructor(storage: Storage) {
		this.storage = storage

		// Create default room
		const defaultRoom = new Room('default', MAP_WIDTH, MAP_HEIGHT)
		this.rooms.set('default', defaultRoom)

		// Place whiteboards near map center
		defaultRoom.addObject({
			id: 'wb-1',
			type: 'whiteboard',
			x: 28 * TILE_SIZE,
			y: 24 * TILE_SIZE,
		})
		defaultRoom.addObject({
			id: 'wb-2',
			type: 'whiteboard',
			x: 34 * TILE_SIZE,
			y: 24 * TILE_SIZE,
		})

		// Place a regional chat zone
		const rcState: RegionalChatState = {
			name: '결계석',
			radius: 80,
			retainHistory: false,
		}
		defaultRoom.addObject({
			id: 'rc-1',
			type: 'regional_chat',
			x: 21 * TILE_SIZE,
			y: 16 * TILE_SIZE,
			state: rcState,
		} as InteractiveObject)
	}

	defaultRoom(): Room | null {
		return this.rooms.get('default') ?? null
	}

	closeAll(): void {
		for (const room of this.rooms.values()) {
			room.closeAll()
		}
	}
}
